keps/0004-add-testing-infrastructure.md

```diff
@@ -68,21 +80,48 @@ This is a two phase proposal.
 
 ### Tests
 
-Add testing packages to:
+KUDO will be tested by several different test suites that validate individual components, KUDO, and frameworks.
+
+#### Unit tests
+
+Unit tests test a single component and require no network connectivity. These should be fast, run on every pull request,
+and test that individual KUDO components behave correctly. Unit tests are written as standard Go tests.
+
+All packages should incorporate unit tests that validate their behavior. Some examples of tests that should be included are:
 
 - pkg/apis - Conformance of objects, serialization, validation
 - pkg/controller - Controllers implement the correct process flows for changes in Instance objects, and creation of PlanExecutions
 - pkg/util - Util functions behave as expected
 
-Integration tests
+Go coverage reports will be used to ensure that changes do not reduce test coverage.
+
+#### Integration tests
+
+Integration tests test the KUDO controller in its entirety. These tests can be run against either a real Kubernetes cluster
+or a local control plane. Initially, all integration tests will be run against every pull request. If this begins to degrade
+velocity, integration tests can be run on master or nightly builds. Integration tests can be written either in Go or using
+the test harness designed in KEP-0008.
+
+The integration tests will consist of a set of representative Frameworks and FrameworkVersions that utilize real world
+KUDO features.
 
-- Identify and build a common set of Frameworks and FrameworkVersions that can demonstrate/validate that functionality of the controller
-  exists in real world situations.
+Integration tests will be hidden behind a Go build tag and will only run when the `integration` tag is specified. Because
+build tags are used to separate integration from unit tests, integration tests written in Go must not be in the same file
+as unit tests.
+
+#### Framework tests
+
+Framework tests test that a framework works correctly. These require a full Kubernetes cluster to run and will be run in CI.
```

Comment:
So we won't test against a local API server stood up by Kubebuilder?  Is it work testing locally there as a (quicker) gate before continuing on to using cloud resources?

Do we forsee a long lived cluster in each environment (AWS, GCP, k3, MKE, OpenShift, etc), or dynamically provisioning each time?

Reply:
> So we won't test against a local API server stood up by Kubebuilder? Is it work testing locally there as a (quicker) gate before continuing on to using cloud resources?

The Framework tests are going to be full end-to-end tests that rely on the framework functioning correctly (something you cannot validate with a local API server only as pods need to start). Local API server tests are documented in the previous section as integration tests that test the KUDO controller (not frameworks).

> Do we forsee a long lived cluster in each environment (AWS, GCP, k3, MKE, OpenShift, etc), or dynamically provisioning each time?

I left this as a bit of an open question below:

>  These clusters can be started either as a part of CI jobs or maintained long term to be used across many CI jobs.

We've got a couple of options:

* Initially we can start them dynamically.
* Once that becomes a problem (maybe it will be from the beginning), we can have long running clusters and namespace our test runs.
* If namespace isolation is not sufficient, maybe we can try having a service that provides ready-built Kubernetes clusters for use in tests.